src/NuGet.Clients/NuGet.SolutionRestoreManager/InfoBarService.cs

```diff
@@ -0,0 +1,165 @@
+
+        public async Task Show(CancellationToken cancellationToken)
+        {
+            await CreateAndShowInfoBarAsync(cancellationToken);
+        }
+
+        public async Task Hide(CancellationToken cancellationToken)
+        {
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            if (!_visible)
+            {
+                return;
+            }
+
+            _closeFromHide = true;
+
+            try
+            {
+                _infoBarUIElement?.Close();
+                _infoBarUIElement = null;
+            }
+            finally
+            {
+                _closeFromHide = false;
+            }
+        }
+
+        protected async Task<IVsInfoBarHost?> GetInfoBarHostAsync(CancellationToken cancellationToken)
+        {
+            IVsUIShell? uiShell = await _asyncServiceProvider.GetServiceAsync<SVsUIShell, IVsUIShell>(throwOnFailure: false);
+            if (uiShell == null)
+            {
+                return null;
+            }
+
+            // Ensure that we are on the UI thread before interacting with the Solution Explorer UI Element
+            await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            if (ErrorHandler.Failed(uiShell.FindToolWindow((uint)__VSFINDTOOLWIN.FTW_fFindFirst, VSConstants.StandardToolWindows.SolutionExplorer, out var windowFrame)))
+            {
+                return null;
+            }
+
+            object tempObject;
+            if (ErrorHandler.Failed(windowFrame.GetProperty((int)__VSFPROPID7.VSFPROPID_InfoBarHost, out tempObject)))
+            {
+                return null;
+            }
+
+            return tempObject as IVsInfoBarHost;
+        }
+
+        protected InfoBarModel GetInfoBarModel()
+        {
+            return new InfoBarModel(
+                new IVsInfoBarTextSpan[] {
+                    new InfoBarTextSpan(Resources.InfoBar_TextMessage),
+                },
+                new IVsInfoBarActionItem[] {
+                    new InfoBarHyperlink(Resources.InfoBar_HyperlinkMessage),
+                },
+                KnownMonikers.StatusWarning);
+        }
+
+        internal async Task CreateAndShowInfoBarAsync(CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            if (_closed || _visible)
+            {
+                return;
+            }
+
+            try
+            {
+                IVsInfoBarUIFactory? infoBarFactory = await _asyncServiceProvider.GetServiceAsync<SVsInfoBarUIFactory, IVsInfoBarUIFactory>(throwOnFailure: false);
+                if (infoBarFactory == null)
+                {
+                    var exception = new NullReferenceException(nameof(infoBarFactory));
+                    await TelemetryUtility.PostFaultAsync(exception, typeof(IVsInfoBarUIFactory).Name, nameof(CreateAndShowInfoBarAsync));
```

Comment:
The class name is supposed to be the caller class, which is this class.
Also, `[CallerMemberName]` should give the name of the method for free without having to explicitly include it.

```suggestion
                    await TelemetryUtility.PostFaultAsync(exception, nameof(InfoBarService));
```